ThunderCloud/ContentController.swift

```diff
@@ -366,37 +366,39 @@ public class ContentController: NSObject {
+    /// - Parameter progressHandler: A closure called when as the download progresses
+    public func downloadFullBundle(buildTimestamp: TimeInterval? = nil, with progressHandler: ContentUpdateProgressHandler?) {
         
-        //Clear existing bundles first
-        if let _currentBundle = bundleDirectory {
-            removeBundle(in: _currentBundle)
-        }
+        removeAllContentBundles()
+        configureBaseURL()
         
-        if let _deltaBundle = deltaDirectory {
-            removeBundle(in: _deltaBundle)
-        }
+        let stormAppId = UserDefaults.standard.string(forKey: "TSCAppId") ?? Storm.API.AppID
         
-        if let _tempDirectory = temporaryUpdateDirectory {
-            removeBundle(in: _tempDirectory)
+        guard let baseString = Storm.API.BaseURL, let version = Storm.API.Version, let appId = stormAppId else {
+            fatalError("Failed to get required parameters to download a storm bundle: missing one of Base URL, Api Version or AppID")
         }
         
-        // Remove intial bundle timestamp as we've now cleared out all evidence of any bundles!
-        initialBundleTimestamp = nil
-        
-        configureBaseURL()
+        guard var urlComponents = URLComponents(string: "\(baseString)/\(version)/apps/\(appId)/bundle") else {
+            fatalError("Failed to create url to download a storm bundle: one of Base URL, Api Version or AppID are invalid url parts")
+        }
+            
+        guard let bundleDirectory = bundleDirectory else {
+            return
+        }
         
-        let stormAppId = UserDefaults.standard.string(forKey: "TSCAppId") ?? Storm.API.AppID
+        if let buildTimestamp = buildTimestamp {
+            urlComponents.queryItems = [
+                URLQueryItem(name: "timestamp", value: "\(Int(buildTimestamp))")
+            ]
+        }
         
-        if let baseString = Storm.API.BaseURL, let version = Storm.API.Version, let appId = stormAppId {
-            
-            if let _fullBundleURL = URL(string: "\(baseString)/\(version)/apps/\(appId)/bundle"), let _destinationURL = bundleDirectory {
-                // We set the initial bundle timestamp here because this bundle will now act as the app's initial bundle!
-                downloadPackage(fromURL: _fullBundleURL, destinationDirectory: _destinationURL, setAsInitialBundle: true, progressHandler: progressHandler)
-            }
+        guard let bundleUrl = urlComponents.url else {
+            fatalError("Failed to create url to download a storm bundle: one of Base URL, Api Version or AppID are invalid url parts")
         }
         
+        // We set the initial bundle timestamp here because this bundle will now act as the app's initial bundle!
+        downloadPackage(fromURL: bundleUrl, destinationDirectory: bundleDirectory, setAsInitialBundle: true, progressHandler: progressHandler)
```

Comment:
If this fails we've deleted all existing bundles?

Reply:
You are correct... luckily the only app that uses this are the GDPC apps which handle all failures really gracefully including on a new app launch ☺️